enhancements/olm/marketplace-api-deprecation.md

```diff
@@ -0,0 +1,130 @@
+
+As of OpenShift 4.4, alternatives for pulling external data have been built into OLM. The appregistry protocol itself is largely replaced by other efforts in OCI, and we cannot continue to rely on it long-term.
+
+### Goals
+
+- Remove support for appregistry-based storage for operator data
+- Remove the on-cluster components that support it
+- Continue to support configuration for OperatorHub
+
+### Non-Goals
+
+- Preserve non-default OperatorSource and CatalogSourceConfig data
+
+## Proposal
+
+### Blocking upgrades in the presence of non-default OperatorSources
+
+The marketplace-operator will be updated to report `Upgradeable: False` if OperatorSources that are not one of the three defaults are detected in the cluster, or any CatalogSourceConfigs.
+
+This will prevent upgrades to clusters that have the marketplace APIs removed until an administrator cleans up the use of appregistry.
+
+The `Upgradeable: False` message should be accompanied by information about the block, with enough pointers to find the resources that are in need of change.
+
+Note: this relies on [this feature of CVO](https://github.com/openshift/cluster-version-operator/pull/291) to be available.
+
+### Removal of the Marketplace APIs
+
+1. Remove the OperatorSource and CatalogSourceConfig CRDs. This will remove all CR instances.
+
+2. Update the default CatalogSources (that are no longer updated by OperatorSources) from `address` type CatalogSources to `image` type CatalogSources, with appropriate poll intervals, pointing the the released catalog images instead of to appregistry repositories. To perform the migration, the OperatorHub Config API will be changed to reconcile `image` CatalogSource defaults instead of OperatorSource defaults.
+
+3. Remove the reconcilation of the OperatorSource and CatalogSourceConfig APIs either by removing the marketplace operator entirely or by removing the relevant portions of it (if it is being retained to manage the OperatorHub API, see below).
+
+### Migration of the OperatorHub Config API
+
+The OperatorHub config API is used to configure which catalogs are available by default in a cluster, and is an important configuration point for disconnected clusters.
+
+This API is currently managed by the marketplace operator and needs to continue to be supported. Currently, the api configures default OperatorSources, and needs to be changed to configure default `image` CatalogSources.
+
+To start with, this will remain in the marketplace-operator. In the future it may be useful to migrate it to the olm.
+
+### Risks and Mitigations
+
+It is possible that some customers are heavily using the appregistry-based catalogs and marketplace APIs. Because we believe this use to be minimal at best, we have not provided an automated migration plan for those cases. Instead, we will block Y upgrades if these apis are in use, with appropriate messages for users.
```

Comment:
We need to have a really good doc that explains how our pipeline migrated from appregistry to bundle images. We are doing the migration they would, we can provide the steps to take IMO.

Reply:
> Otherwise, a manual guide for migrating custom catalogs should be straightforward to follow.

+1 sorry.